pkg/operator/ceph/object/cosi/controller.go

```diff
@@ -0,0 +1,291 @@
+
+	// check whether cosi controller is running
+	cosiControllerDeployment := &appsv1.Deployment{}
+	err := r.client.Get(r.opManagerContext, types.NamespacedName{Name: CosiControllerDeployment, Namespace: CosiControllerNamespace}, cosiControllerDeployment)
+	if err != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to get cosi controller deployment. %+v", err)
+	}
+
+	// Fetch the CephCOSIDriver instance
+	err = r.client.Get(r.opManagerContext, request.NamespacedName, cephCOSIDriver)
+	if err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			logger.Info("CephCOSIDriver resource not found. Starting with Default values")
+		} else {
+			// Error reading the object - requeue the request.
+			return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to get CephCOSIDriver. %+v", err)
+		}
+	}
+
+	// Set the default CephCOSIDriver name if not already set
+	if cephCOSIDriver.Name == "" {
+		cephCOSIDriver.Name = DefaultCephCOSIDriverName
+	}
+
+	// Set the default CephCOSIDriver namespace if not already set
+	if cephCOSIDriver.Namespace == "" {
+		cephCOSIDriver.Namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	}
+
+	// TODO address non default use cases
+
+	// Check whether object store is running
+	list, err := r.context.RookClientset.CephV1().CephObjectStores(cephCOSIDriver.Namespace).List(r.opManagerContext, metav1.ListOptions{})
+	if err != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to list object stores. %+v", err)
+	}
+
+	// Check for object store part of Multisite
+	if len(list.Items) == 0 {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("no object store found, requeuing")
+	}
+
+	// TODO create CephObjectStoreUser name cosi for each object store, if not already created
+
+	// Start the Ceph COSI Driver
+	err = r.startCephCOSIDriver(cephCOSIDriver)
+	if err != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to start Ceph COSI Driver. %+v", err)
+	}
+
+	return reconcile.Result{}, *cephCOSIDriver, nil
+}
+
+// Start the Ceph COSI Driver
+func (r *ReconcileCephCOSIDriver) startCephCOSIDriver(cephCOSIDriver *cephv1.CephCOSIDriver) error {
+	// Create the Ceph COSI Driver Deployment
+	cephcosidriverDeployment, err := createCephCOSIDriverDeployment(cephCOSIDriver)
+	if err != nil {
+		return fmt.Errorf("failed to create Ceph COSI Driver deployment. %+v", err)
+	}
+
+	_, err = r.context.Clientset.AppsV1().Deployments(cephCOSIDriver.Namespace).Create(r.opManagerContext, cephcosidriverDeployment, metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			// TODO updates
+			return fmt.Errorf("failed to create Ceph COSI Driver deployment. %+v", err)
+		}
+		logger.Infof("Ceph COSI Driver deployment already exists")
+	} else {
+		logger.Infof("Ceph COSI Driver deployment started")
+	}
+
+	return nil
+}
+
+func createCephCOSIDriverDeployment(cephCOSIDriver *cephv1.CephCOSIDriver) (*appsv1.Deployment, error) {
```

Comment:
Let's factor out these methods into a spec.go, so the controller.go only contains things specific to the controller. This makes it more clear for unit tests, and is a common pattern in rook.